### PR TITLE
raftstore: fix find_propose_time may return older instant

### LINKED ISSUE #9748: The propose time may be smaller than the real one
### What version of TiKV are you using?
master

### What operating system and CPU are you using?
doesn't matter

### Steps to reproduce
doesn't matter

### What did you expect?
right propose time

### What did happened?
https://github.com/tikv/tikv/blob/311515d67331dee4bf1f34ebeb6200bd6061f711/components/raftstore/src/store/peer.rs#L104-L107
If the `idx` is got from `back.binary_search_by_key(&key, map)`, the `self.queue[idx]` is not the one we want.
The underlying data structure of `VecDeque` is a ring buffer. So if it's the back slice, the `idx` should add the length of front slice.
As an alternative, we can use the `VecDeque's binary_search_by_key` function though it's not stable yet.

Luckily, this bug can not lead to serious consequences because it may only return the smaller `propose time`.
As a result, it may cause a smaller `commit log duration` and `maybe_renew_leader_lease` will use this smaller one to update lease which doesn't break correctness.

### PR DESCRIPTION
**Pull request by NingLin-P**
Signed-off-by: linning <linningde25@gmail.com>

### What problem does this PR solve?

Issue Number: close #9748 

Problem Summary:

See https://github.com/tikv/tikv/issues/9748

### What is changed and how it works?

What's Changed:

Use `VecDeque::binary_search_by_key` to find proposal

### Check List 

Tests 

- Unit test

### Release note 
- raftstore: fix find_propose_time may return older instant

### CODE REVIEW

**Comment by NingLin-P** [comment]
This PR also add test to cover the case, I had manually check that the test will fail in the master branch and pass in this PR.

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -3965,10 +3963,24 @@ mod tests {
...
                 must_pass_epoch_check: false,
             });
+        };
+        for index in 1..=100 {
+            push_proposal(&mut pq, index);
         }
         let mut pre_remove = 0;
-        for remove_i in &[0, 1, 65, 98, 100] {
-            let remove_i = *remove_i;
+        for remove_i in 1..=100 {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:3971]: How long does the case take?

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -3955,8 +3953,8 @@ mod tests {
             ProposalQueue::new("tag".to_owned());
         let t = monotonic_raw_now();
         let gen_term = |index: u64| (index / 10) + 1;
-        for index in 1..=100 {
-            let renew_lease_time = if index % 3 == 1 { None } else { Some(t) };
+        let push_proposal = |pq: &mut ProposalQueue<_>, index: u64| {
+            let renew_lease_time = if index % 3 == 0 { None } else { Some(t) };
```
Comment (gengliqi) [inline components/raftstore/src/store/peer.rs:3957]: How about using different values?

File: components/raftstore/src/store/peer.rs (hunk 1 above)
Comment (NingLin-P) [reply components/raftstore/src/store/peer.rs:3971]: There is no time-consuming operation, the test will finish before a human can notice.

File: components/raftstore/src/store/peer.rs (hunk 2 above)
Comment (NingLin-P) [reply components/raftstore/src/store/peer.rs]: I use `u64_to_timespec` to get a unique propose time for each proposal instead, PTAL

**Review by BusyJay — APPROVED** [review]

**Review by gengliqi — APPROVED** [review]

**Pull request merged**